cudax/include/cuda/experimental/__utility/basic_any/dynamic_any_cast.cuh

```diff
@@ -0,0 +1,113 @@
+
+_CCCL_PUSH_MACROS
+#undef interface
+
+namespace cuda::experimental
+{
+//!
+//! __dynamic_any_cast_fn
+//!
+template <class _DstInterface>
+struct __dynamic_any_cast_fn
+{
+  //! @throws bad_any_cast when \c __src cannot be dynamically cast to a
+  //! \c basic_any<_DstInterface>.
+  _CCCL_TEMPLATE(class _SrcInterface)
+  _CCCL_REQUIRES(__any_castable_to<basic_any<_SrcInterface>, basic_any<_DstInterface>>)
+  _CCCL_NODISCARD _CUDAX_HOST_API auto operator()(basic_any<_SrcInterface>&& __src) const -> basic_any<_DstInterface>
+  {
+    auto __dst = __basic_any_access::__make<_DstInterface>();
+    __basic_any_access::__cast_to(_CUDA_VSTD::move(__src), __dst);
+    return __dst;
+  }
+
+  //! @throws bad_any_cast when \c __src cannot be dynamically cast to a
+  //! \c basic_any<_DstInterface>.
+  _CCCL_TEMPLATE(class _SrcInterface)
+  _CCCL_REQUIRES(__any_castable_to<basic_any<_SrcInterface>&, basic_any<_DstInterface>>)
+  _CCCL_NODISCARD _CUDAX_HOST_API auto operator()(basic_any<_SrcInterface>& __src) const -> basic_any<_DstInterface>
+  {
+    auto __dst = __basic_any_access::__make<_DstInterface>();
+    __basic_any_access::__cast_to(__src, __dst);
+    return __dst;
+  }
+
+  //! @throws bad_any_cast when \c __src cannot be dynamically cast to a
+  //! \c basic_any<_DstInterface>.
+  _CCCL_TEMPLATE(class _SrcInterface)
+  _CCCL_REQUIRES(__any_castable_to<basic_any<_SrcInterface> const&, basic_any<_DstInterface>>)
+  _CCCL_NODISCARD _CUDAX_HOST_API auto
+  operator()(basic_any<_SrcInterface> const& __src) const -> basic_any<_DstInterface>
+  {
+    auto __dst = __basic_any_access::__make<_DstInterface>();
+    __basic_any_access::__cast_to(__src, __dst);
+    return __dst;
+  }
+
+  //! @returns \c nullptr when \c __src cannot be dynamically cast to a
+  //! \c basic_any<_DstInterface>.
+  _CCCL_TEMPLATE(class _SrcInterface)
+  _CCCL_REQUIRES(__any_castable_to<basic_any<_SrcInterface>*, basic_any<_DstInterface>>)
+  _CCCL_NODISCARD _CUDAX_HOST_API auto operator()(basic_any<_SrcInterface>* __src) const -> basic_any<_DstInterface>
+  {
+    auto __dst = __basic_any_access::__make<_DstInterface>();
+    __basic_any_access::__cast_to(__src, __dst);
+    return __dst;
+  }
+
+  //! @returns \c nullptr when \c __src cannot be dynamically cast to a
+  //! \c basic_any<_DstInterface>.
+  _CCCL_TEMPLATE(class _SrcInterface)
+  _CCCL_REQUIRES(__any_castable_to<basic_any<_SrcInterface> const*, basic_any<_DstInterface>>)
+  _CCCL_NODISCARD _CUDAX_HOST_API auto
+  operator()(basic_any<_SrcInterface> const* __src) const -> basic_any<_DstInterface>
+  {
+    auto __dst = __basic_any_access::__make<_DstInterface>();
+    __basic_any_access::__cast_to(__src, __dst);
+    return __dst;
+  }
+};
+
+//!
+//! dynamic_any_cast
+//!
+//! @throws bad_any_cast when \c from cannot be dynamically cast to a
+//! \c basic_any<_DstInterface>
+//!
+template <class _DstInterface>
+inline constexpr __dynamic_any_cast_fn<_DstInterface> dynamic_any_cast{};
```

Comment:
I believe this needs to be
```suggestion
_CCCL_GLOBAL_CONSTANT __dynamic_any_cast_fn<_DstInterface> dynamic_any_cast{};
```
because MSVC really doesnt like gobal variable templates